### PR TITLE
[Notebook port 4835] Add UNIX socket support to notebook server

### PR DESCRIPTION
**Pull request by jtpio**
This is a port of [jupyter/notebook#4835](https://github.com/jupyter/notebook/pull/4835). (The changes could not be easily cherry-picked because it requires substantial changes when porting the tests.)

Overview of the changes:

* enables connection to a jupyter server via UNIX sockets.
* adds integration tests for launching, listing, and shutting-down UNIX socket servers.
* runs the integration tests in a new, separate Github workflow.
* adds a `--integration-tests` flag to the pytest CLI options for running the integration tests. Otherwise, these tests are ignored. (They shouldn't _really_ be tested locally).
* improves, adds, and tests the handling of some public properties related to the ServerApp's URLs.
* adds some utility functions for making sync/async HTTP **or** UNIX socket requests to a running Jupyter server.

### CODE REVIEW

**Comment by kevin-bates** [comment]
I went ahead and dismissed the two remaining codeQL issues since they were only in test code (side effect stuff).

**Review by Zsailer — APPROVED** [review]
Thanks, @jtpio! :rocket: This looks good to me. 

@kevin-bates or @mwakaba2, would either of you be willing to do a quick scan of the code to make sure nothing looks out of the ordinary? After one of you approves, I'll go ahead a merge and release. Thank you!

**Review by kevin-bates — APPROVED** [review]
These changes look good.  Not sure if the dismissed codeQL issues in the test code will "come back to life" or not, but let's see. :smile:

File: setup.cfg [hunk 1]
```diff
@@ -42,6 +42,7 @@ install_requires =
     prometheus_client
     anyio>=3.1.0,<4
     websocket-client
+    requests-unixsocket
```
Comment (lazka) [inline setup.cfg:45]: Is this actually used somewhere?

**Pull request merged**